compiler/src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -712,50 +719,72 @@ class Inliner(call: tpd.Tree, rhsToInline: tpd.Tree)(implicit ctx: Context) {
-                    t.symbol.asType :: syms
-                  case _ =>
-                    syms
-                }
-                foldOver(syms1, t)
+        def getBoundVarsList(pat: Tree, tpt: Tree): List[(TypeSymbol, Boolean)] = {
+          // UnApply nodes with pattern bound variables translate to something like this
+          //   UnApply[t @ t](pats)(implicits): T[t]
+          // Need to traverse any binds in type arguments of the UnAppyl to get the set of
+          // all instantiable type variables. Test case is pos/inline-caseclass.scala.
+          val allTpts = tpt :: (pat match {
+            case UnApply(TypeApply(_, tpts), _, _) => tpts
+            case _ => Nil
+          })
+
+          val getBinds = new TreeAccumulator[Set[TypeSymbol]] {
+            def apply(syms: Set[TypeSymbol], t: Tree)(implicit ctx: Context): Set[TypeSymbol] = {
+              val syms1 = t match {
+                case t: Bind if t.symbol.isType =>
+                  syms + t.symbol.asType
+                case _ => syms
               }
+              foldOver(syms1, t)
             }
-            var boundVars = getBoundVars(Nil, tpt)
-            // UnApply nodes with pattern bound variables translate to something like this
-            //   UnApply[t @ t](pats)(implicits): T[t]
-            // Need to traverse any binds in type arguments of the UnAppyl to get the set of
-            // all instantiable type variables. Test case is pos/inline-caseclass.scala.
-            pat1 match {
-              case UnApply(TypeApply(_, tpts), _, _) =>
-                for (tpt <- tpts) boundVars = getBoundVars(boundVars, tpt)
-              case _ =>
-            }
-            for (bv <- boundVars) {
-              val TypeBounds(lo, hi) = bv.info.bounds
-              ctx.gadt.addBound(bv, lo, isUpper = false)
-              ctx.gadt.addBound(bv, hi, isUpper = true)
+          }
+          val binds = allTpts.foldLeft[Set[TypeSymbol]](Set.empty)(getBinds.apply(_, _))
+          val findBoundVars = new TypeAccumulator[List[(TypeSymbol, Boolean)]] {
+            def apply(syms: List[(TypeSymbol, Boolean)], t: Type) = {
+              val syms1 = t match {
+                 case tr: TypeRef if tr.symbol.is(Case) && binds.contains(tr.symbol.asType) =>
+                  (tr.typeSymbol.asType, variance >= 0) :: syms
+                case _ =>
+                  syms
+              }
+              foldOver(syms1, t)
             }
+          }
+          findBoundVars(Nil, tpt.tpe)
+        }
+
+        def registerAsGadtSyms(list: List[(TypeSymbol, Boolean)])(implicit ctx: Context): Unit =
+          list.foreach { case (sym, _) =>
+            val TypeBounds(lo, hi) = sym.info.bounds
+            ctx.gadt.addBound(sym, lo, isUpper = false)
+            ctx.gadt.addBound(sym, hi, isUpper = true)
+          }
+
+        def addTypeBindings(list: List[(TypeSymbol, Boolean)])(implicit ctx: Context): Unit =
```

Comment:
I've taken another look at my approach - it looks like currently it's really _not_ possible to get multiple bindings in a list, but relying on that could be fragile and it's relatively obvious that we should do the same as `IsFullyDefinedAccumulator`, so I've reworked the `List` into a `type TypeBindsMap = SimpleIdentityMap[TypeSymbol, java.lang.Boolean]` and added appropriate value-merging logic to `findBoundVars`.

Unless I'm missing something, we cannot merge `findBoundVars` and `getBinds`, since the latter locates symbols bound by the current pattern (as opposed to enclosing patterns) and the former find the variance of the symbol by traversing the type. I've renamed `findBoundVars` to `extractBindVariance` and added a comment to make this clearer.